### PR TITLE
Refactor CFeeRate to use FeeFrac internally

### PR DESCRIPTION
**Pull request by polespinasa**
This project will fix precision issues in the CFeeRate class. The current workaround is to replace CFeeRate with FeeFrac when precision problems occur. A better long-term solution is to improve CFeeRate so it handles precision correctly and does not need to be replaced.
The main task is to refactor CFeeRate to use FeePerVSize internally. This should fix the precision issue and avoid changes across the rest of the codebase. 

### Expected outcomes

- [x] Refactor CFeeRate to wrap FeePerVSize while maintaining backward compatibility.
- [x] Ensure precise fee rate calculations with no loss of accuracy.
- [x] Add targeted unit tests to verify edge cases where CFeeRate previously lost precision.
- [x] Ensure all CI and fuzz tests pass after changes.

TODO:
- ~Fix the serialization issue~

### CODE REVIEW

**Comment by polespinasa** [comment]
On test `amount_tests` we have some failing tests:
```c++
    // lost precision (can only resolve satoshis per kB)
    BOOST_CHECK(CFeeRate(CAmount(1), 1001) == CFeeRate(0));
    BOOST_CHECK(CFeeRate(CAmount(2), 1001) == CFeeRate(1));
    // some more integer checks
    BOOST_CHECK(CFeeRate(CAmount(26), 789) == CFeeRate(32));
    BOOST_CHECK(CFeeRate(CAmount(27), 789) == CFeeRate(34));
```
I think is because of decimal precision, how should we handle that?
For example for the first case the comparison is `1/1001 != 0/1000` because we have `0.00009 != 0`

**Review by ismaelsadeeq — COMMENTED** [review]
Did a first pass today

I think the current implementation of the `CFeeRate` addition operator has issues. 

Reminder:
a/b + c/d ≠ (a + b) / (c + d)
The correct formula is:
a/b + c/d = (ad + bc) / bd

You should either implement this directly or evaluate the fractional fee rates and sum them. That’s the approach I took here.

Also, since `CAmount` no longer loses precision, you’ll need to update the related unit tests accordingly. In some cases, comparing the evaluated fraction is now necessary because we are no longer storing the evaluated fraction.

See the diff below for an implementation that passes unit tests.

<details>
<summary>see diff
</summary>

```diff
diff --git a/src/policy/feerate.cpp b/src/policy/feerate.cpp
index eb0cba5c67a..7920b8bcec5 100644
--- a/src/policy/feerate.cpp
+++ b/src/policy/feerate.cpp
@@ -9,37 +9,32 @@
 
 #include <cmath>
 
-CFeeRate::CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes)
+CFeeRate::CFeeRate(const CAmount& nFeePaid, int32_t num_bytes)
 {
-    const int64_t nSize{num_bytes};
-
-    if (nSize > 0) {
-        nSatoshisPerK = nFeePaid * 1000 / nSize;
-    } else {
-        nSatoshisPerK = 0;
+    if (num_bytes > 0) {
+        fee_rate = FeePerVSize(nFeePaid, num_bytes);
     }
 }
 
-CAmount CFeeRate::GetFee(uint32_t num_bytes) const
+CAmount CFeeRate::GetFee(int32_t num_bytes) const
 {
-    const int64_t nSize{num_bytes};
-
-    // Be explicit that we're converting from a double to int64_t (CAmount) here.
-    // We've previously had issues with the silent double->int64_t conversion.
-    CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
-
-    if (nFee == 0 && nSize != 0) {
-        if (nSatoshisPerK > 0) nFee = CAmount(1);
-        if (nSatoshisPerK < 0) nFee = CAmount(-1);
+    auto fee = fee_rate.EvaluateFeeUp(num_bytes);
+    if (fee == 0  && num_bytes != 0)
+    {
+        if (GetFeePerK() > 0) fee = CAmount(1);
+        if (GetFeePerK() < 0) fee = CAmount(-1);
     }
-
-    return nFee;
+    return fee;
 }
 
 std::string CFeeRate::ToString(const FeeEstimateMode& fee_estimate_mode) const
 {
+    const CAmount rate_per_kvb = fee_rate.fee * 1000 / fee_rate.size;
+
     switch (fee_estimate_mode) {
-    case FeeEstimateMode::SAT_VB: return strprintf("%d.%03d %s/vB", nSatoshisPerK / 1000, nSatoshisPerK % 1000, CURRENCY_ATOM);
-    default:                      return strprintf("%d.%08d %s/kvB", nSatoshisPerK / COIN, nSatoshisPerK % COIN, CURRENCY_UNIT);
+    case FeeEstimateMode::SAT_VB:
+        return strprintf("%d.%03d %s/vB", rate_per_kvb / 1000, rate_per_kvb % 1000, CURRENCY_ATOM);
+    default:
+        return strprintf("%d.%08d %s/kvB", rate_per_kvb / COIN, rate_per_kvb % COIN, CURRENCY_UNIT);
     }
 }
diff --git a/src/policy/feerate.h b/src/policy/feerate.h
index d742a43acc7..044bfd0ba2d 100644
--- a/src/policy/feerate.h
+++ b/src/policy/feerate.h
@@ -6,6 +6,7 @@
 #ifndef BITCOIN_POLICY_FEERATE_H
 #define BITCOIN_POLICY_FEERATE_H
 
+#include <util/feefrac.h>
 #include <consensus/amount.h>
 #include <serialize.h>
 
@@ -17,6 +18,8 @@
 const std::string CURRENCY_UNIT = "BTC"; // One formatted unit
 const std::string CURRENCY_ATOM = "sat"; // One indivisible minimum value unit
 
+const int16_t KiloVirtualByte = 1000;
+
 /* Used to determine type of fee estimation requested */
 enum class FeeEstimateMode {
     UNSET,        //!< Use default settings based on other criteria
@@ -27,20 +30,21 @@ enum class FeeEstimateMode {
 };
 
 /**
- * Fee rate in satoshis per kilovirtualbyte: CAmount / kvB
+ * Fee rate in satoshis per virtualbyte: CAmount / vB
+ * Wrapping FeeFrac to handle
  */
 class CFeeRate
 {
 private:
-    /** Fee rate in sat/kvB (satoshis per 1000 virtualbytes) */
-    CAmount nSatoshisPerK;
+    /** Fee rate in sat/vB (satoshis per virtualbyte) */
+    FeePerVSize fee_rate;
 
 public:
-    /** Fee rate of 0 satoshis per kvB */
-    CFeeRate() : nSatoshisPerK(0) { }
+    /** Fee rate of 0 satoshis per vB */
+    CFeeRate() = default;
+    /** Initialize a Fee rate of  _nSatoshisPerK satoshis per 1kvB */
     template<std::integral I> // Disallow silent float -> int conversion
-    explicit CFeeRate(const I _nSatoshisPerK): nSatoshisPerK(_nSatoshisPerK) {
-    }
+    explicit CFeeRate(const I _nSatoshisPerK) : fee_rate(FeePerVSize(_nSatoshisPerK, KiloVirtualByte)) {}
 
     /**
      * Construct a fee rate from a fee in satoshis and a vsize in vB.
@@ -48,31 +52,48 @@ public:
      * param@[in]   nFeePaid    The fee paid by a transaction, in satoshis
      * param@[in]   num_bytes   The vsize of a transaction, in vbytes
      */
-    CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes);
+    CFeeRate(const CAmount& nFeePaid, int32_t num_bytes);
 
     /**
      * Return the fee in satoshis for the given vsize in vbytes.
      * If the calculated fee would have fractional satoshis, then the
      * returned fee will always be rounded up to the nearest satoshi.
      */
-    CAmount GetFee(uint32_t num_bytes) const;
+    CAmount GetFee(int32_t num_bytes) const;
 
     /**
      * Return the fee in satoshis for a vsize of 1000 vbytes
      */
-    CAmount GetFeePerK() const { return nSatoshisPerK; }
-    friend bool operator<(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK < b.nSatoshisPerK; }
-    friend bool operator>(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK > b.nSatoshisPerK; }
-    friend bool operator==(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK == b.nSatoshisPerK; }
-    friend bool operator<=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK <= b.nSatoshisPerK; }
-    friend bool operator>=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK >= b.nSatoshisPerK; }
-    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK != b.nSatoshisPerK; }
-    CFeeRate& operator+=(const CFeeRate& a) { nSatoshisPerK += a.nSatoshisPerK; return *this; }
+    CAmount GetFeePerK() const {
+        return (CAmount)((fee_rate.fee * KiloVirtualByte) / fee_rate.size);
+    }
+    /**
+     * Return the fee in satoshis for a vsize of 1 vbytes
+     */
+    CAmount GetFeePerVb() const {
+        return (CAmount)(fee_rate.fee / fee_rate.size);
+    }
+    friend bool operator<(const CFeeRate& a, const CFeeRate& b) { return a.fee_rate << b.fee_rate; }
+    friend bool operator>(const CFeeRate& a, const CFeeRate& b) { return a.fee_rate >> b.fee_rate; }
+    friend bool operator==(const CFeeRate& a, const CFeeRate& b) { return FeeRateCompare(a.fee_rate, b.fee_rate) == std::weak_ordering::equivalent; }
+    friend bool operator<=(const CFeeRate& a, const CFeeRate& b) {
+        auto compare = FeeRateCompare(a.fee_rate, b.fee_rate);
+        return compare == std::weak_ordering::equivalent || compare == std::weak_ordering::less;
+    }
+    friend bool operator>=(const CFeeRate& a, const CFeeRate& b) {
+        auto compare = FeeRateCompare(a.fee_rate, b.fee_rate);
+        return compare == std::weak_ordering::equivalent || compare == std::weak_ordering::greater;
+    }
+    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return FeeRateCompare(a.fee_rate, b.fee_rate) != std::weak_ordering::equivalent; }
+    CFeeRate& operator+=(const CFeeRate& a) {
+        fee_rate  = FeePerVSize(GetFeePerK() + a.GetFeePerK(), KiloVirtualByte);
+        return *this;
+    }
     std::string ToString(const FeeEstimateMode& fee_estimate_mode = FeeEstimateMode::BTC_KVB) const;
-    friend CFeeRate operator*(const CFeeRate& f, int a) { return CFeeRate(a * f.nSatoshisPerK); }
-    friend CFeeRate operator*(int a, const CFeeRate& f) { return CFeeRate(a * f.nSatoshisPerK); }
+    friend CFeeRate operator*(const CFeeRate& f, int a) { return CFeeRate(a * f.fee_rate.fee, f.fee_rate.size); }
+    friend CFeeRate operator*(int a, const CFeeRate& f) { return CFeeRate(a * f.fee_rate.fee, f.fee_rate.size); }
 
-    SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.nSatoshisPerK); }
+    SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.fee_rate); }
 };
 
 #endif // BITCOIN_POLICY_FEERATE_H
diff --git a/src/test/amount_tests.cpp b/src/test/amount_tests.cpp
index e5ab1cfb902..b3f9af37fc5 100644
--- a/src/test/amount_tests.cpp
+++ b/src/test/amount_tests.cpp
@@ -77,14 +77,14 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
     BOOST_CHECK(CFeeRate(CAmount(-1), 1000) == CFeeRate(-1));
     BOOST_CHECK(CFeeRate(CAmount(0), 1000) == CFeeRate(0));
     BOOST_CHECK(CFeeRate(CAmount(1), 1000) == CFeeRate(1));
-    // lost precision (can only resolve satoshis per kB)
-    BOOST_CHECK(CFeeRate(CAmount(1), 1001) == CFeeRate(0));
-    BOOST_CHECK(CFeeRate(CAmount(2), 1001) == CFeeRate(1));
+    // we don't lose precision
+    BOOST_CHECK(CFeeRate(CAmount(1), 1001) > CFeeRate(0));
+    BOOST_CHECK(CFeeRate(CAmount(2), 1001) > CFeeRate(1));
     // some more integer checks
-    BOOST_CHECK(CFeeRate(CAmount(26), 789) == CFeeRate(32));
-    BOOST_CHECK(CFeeRate(CAmount(27), 789) == CFeeRate(34));
+    BOOST_CHECK(CFeeRate(CAmount(26), 789).GetFeePerK() == CFeeRate(32).GetFeePerK());
+    BOOST_CHECK(CFeeRate(CAmount(27), 789).GetFeePerK() == CFeeRate(34).GetFeePerK());
     // Maximum size in bytes, should not crash
-    CFeeRate(MAX_MONEY, std::numeric_limits<uint32_t>::max()).GetFeePerK();
+    CFeeRate(MAX_MONEY, std::numeric_limits<int32_t>::max()).GetFeePerK();
 
     // check multiplication operator
     // check multiplying by zero
diff --git a/src/validation.cpp b/src/validation.cpp
index 5ad2ebdcd7e..298528aa00a 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1396,7 +1396,7 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
         auto iter = m_pool.GetIter(ws.m_ptx->GetHash());
         Assume(iter.has_value());
         const auto effective_feerate = args.m_package_feerates ? ws.m_package_feerate :
-            CFeeRate{ws.m_modified_fees, static_cast<uint32_t>(ws.m_vsize)};
+            CFeeRate{ws.m_modified_fees, static_cast<int32_t>(ws.m_vsize)};
         const auto effective_feerate_wtxids = args.m_package_feerates ? all_package_wtxids :
             std::vector<Wtxid>{ws.m_ptx->GetWitnessHash()};
         results.emplace(ws.m_ptx->GetWitnessHash(),
@@ -1460,7 +1460,7 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
 
     if (!ConsensusScriptChecks(args, ws)) return MempoolAcceptResult::Failure(ws.m_state);
 
-    const CFeeRate effective_feerate{ws.m_modified_fees, static_cast<uint32_t>(ws.m_vsize)};
+    const CFeeRate effective_feerate{ws.m_modified_fees, static_cast<int32_t>(ws.m_vsize)};
     // Tx was accepted, but not added
     if (args.m_test_accept) {
         return MempoolAcceptResult::Success(std::move(m_subpackage.m_replaced_transactions), ws.m_vsize,
@@ -1612,7 +1612,7 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
         }
         if (args.m_test_accept) {
             const auto effective_feerate = args.m_package_feerates ? ws.m_package_feerate :
-                CFeeRate{ws.m_modified_fees, static_cast<uint32_t>(ws.m_vsize)};
+                CFeeRate{ws.m_modified_fees, static_cast<int32_t>(ws.m_vsize)};
             const auto effective_feerate_wtxids = args.m_package_feerates ? all_package_wtxids :
                 std::vector<Wtxid>{ws.m_ptx->GetWitnessHash()};
             results.emplace(ws.m_ptx->GetWitnessHash(),
```

</details>

TODO for you.

- Run the unit tests to confirm correctness.
- Run fuzz tests and ensure there are no failures.
- Run linters and check formatting across the diff.
- Fix serialization issue comment below

**Comment by polespinasa** [comment]
Is there any reason in these two functions why you prefer doing the conversion manually instead of relaying in `EvaluateFeeUp()` function?

> ```diff
> +    CAmount GetFeePerK() const {
> +        return (CAmount)((fee_rate.fee * KiloVirtualByte) / fee_rate.size);
> +    }
> +    /**
> +     * Return the fee in satoshis for a vsize of 1 vbytes
> +     */
> +    CAmount GetFeePerVb() const {
> +        return (CAmount)(fee_rate.fee / fee_rate.size);
> +    }
> ```


I think this approach can fail if `nSatoshisPerK->size` is 0, see copmment in feefrac.h:L197
```c++
* Requires this->size > 0, at_size >= 0, and that the correct result fits in a int64_t. This
* is guaranteed to be the case when 0 <= at_size <= this->size.
```
> ```diff
> -CAmount CFeeRate::GetFee(uint32_t num_bytes) const
> +CAmount CFeeRate::GetFee(int32_t num_bytes) const
>  {
> -    const int64_t nSize{num_bytes};
> -
> -    // Be explicit that we're converting from a double to int64_t (CAmount) here.
> -    // We've previously had issues with the silent double->int64_t conversion.
> -    CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
> -
> -    if (nFee == 0 && nSize != 0) {
> -        if (nSatoshisPerK > 0) nFee = CAmount(1);
> -        if (nSatoshisPerK < 0) nFee = CAmount(-1);
> +    auto fee = fee_rate.EvaluateFeeUp(num_bytes);
> +    if (fee == 0  && num_bytes != 0)
> +    {
> +        if (GetFeePerK() > 0) fee = CAmount(1);
> +        if (GetFeePerK() < 0) fee = CAmount(-1);
>      }
> -
> -    return nFee;
> +    return fee;
>  }
> ```

**Review by polespinasa — COMMENTED** [review]
0c20b8353988728b7c195efec14a7e83304154b8 fixes a functional test by increasing by 0.(...)1 the `estimated_feerate`

File: src/policy/feerate.h [hunk 1]
```diff
@@ -60,19 +61,19 @@ class CFeeRate
...
+    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerV != b.nSatoshisPerV; }
+    CFeeRate& operator+=(const CFeeRate& a) { nSatoshisPerV += a.nSatoshisPerV; return *this; }
     std::string ToString(const FeeEstimateMode& fee_estimate_mode = FeeEstimateMode::BTC_KVB) const;
-    friend CFeeRate operator*(const CFeeRate& f, int a) { return CFeeRate(a * f.nSatoshisPerK); }
-    friend CFeeRate operator*(int a, const CFeeRate& f) { return CFeeRate(a * f.nSatoshisPerK); }
+    friend CFeeRate operator*(const CFeeRate& f, int a) { return CFeeRate(a * f.nSatoshisPerV.fee); }
+    friend CFeeRate operator*(int a, const CFeeRate& f) { return CFeeRate(a * f.nSatoshisPerV.fee); }
 
-    SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.nSatoshisPerK); }
+    SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.nSatoshisPerV); }
```
[the remark this answers is not in the document]
Comment (polespinasa) [reply src/policy/feerate.h]: Tried to fix in f7b67ddca4a8514005caa9a9bf3290b79d24bf1f

I'm not sure if using `fee` and `size` separated is a good approach but is the only one that compiled.
I've left one of the other approaches commented but that doesn't work as you cannot override the object with `=`.

File: src/policy/feerate.cpp [hunk 2]
```diff
@@ -9,37 +9,36 @@
 
 #include <cmath>
 
-CFeeRate::CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes)
+CFeeRate::CFeeRate(const CAmount& nFeePaid, int32_t num_bytes)
 {
-    const int64_t nSize{num_bytes};
-
-    if (nSize > 0) {
-        nSatoshisPerK = nFeePaid * 1000 / nSize;
```
Comment (ismaelsadeeq) [inline src/policy/feerate.cpp:17]: I think the reason why your functional tests are failing is because here we are not rounding up (It is implicitly being round down).

But in your implementation of `GetFeePerK` you round up.

File: src/policy/feerate.h [hunk 3]
```diff
@@ -27,52 +28,76 @@ enum class FeeEstimateMode {
...
      */
-    CAmount GetFeePerK() const { return nSatoshisPerK; }
-    friend bool operator<(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK < b.nSatoshisPerK; }
-    friend bool operator>(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK > b.nSatoshisPerK; }
-    friend bool operator==(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK == b.nSatoshisPerK; }
-    friend bool operator<=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK <= b.nSatoshisPerK; }
-    friend bool operator>=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK >= b.nSatoshisPerK; }
-    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK != b.nSatoshisPerK; }
-    CFeeRate& operator+=(const CFeeRate& a) { nSatoshisPerK += a.nSatoshisPerK; return *this; }
+    CAmount GetFeePerK() const { return CAmount(nSatoshisPerV.EvaluateFeeUp(1000)); }
```
Comment (ismaelsadeeq) [inline src/policy/feerate.h:64]:
```suggestion
CAmount GetFeePerK() const { return CAmount(nSatoshisPerV.EvaluateFeeDown(1000)); }
```

File: src/policy/feerate.cpp (hunk 2 above)
Comment (polespinasa) [reply src/policy/feerate.cpp:17]: Great catch! That was it!

File: src/policy/feerate.h [hunk 4]
```diff
@@ -27,52 +28,76 @@ enum class FeeEstimateMode {
...
- * Fee rate in satoshis per kilovirtualbyte: CAmount / kvB
+ * Fee rate in satoshis per virtualbyte: CAmount / vB
+ * Wrapping FeeFrac to handle precision issues
  */
 class CFeeRate
 {
 private:
-    /** Fee rate in sat/kvB (satoshis per 1000 virtualbytes) */
-    CAmount nSatoshisPerK;
+    /** Fee rate in sat/kvB (satoshis per 1 virtualbyte) */
```
[the remark this answers is not in the document]
Comment (polespinasa) [reply src/policy/feerate.h]: Actually changed to sats/vB and specified (satoshis per N virtualbytes) as we're using FeeFrac and we can use different proportions

File: src/policy/feerate.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (polespinasa) [reply src/policy/feerate.h]: > I have not attempted but I think it should

You mean that the approach of having fee and size separated is a good approach? Or that the other approach should work?

**Comment by polespinasa** [comment]
9295d57296b47461b79f705f7364e7de39b824e8 cleans code on serialization part.

**Comment by polespinasa** [comment]
rebase on top of master de3dee5c50663d33aea36db9eae9cdb16265a30a

**Comment by polespinasa** [comment]
A PR was open to Bitcoin Core https://github.com/bitcoin/bitcoin/pull/32750
